analyzers/src/SonarAnalyzer.CSharp/Rules/SillyMathematicalComparison.cs

```diff
@@ -0,0 +1,108 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class SillyMathematicalComparison : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S2198";
+        private const string MessageFormat = "Comparison to integral constant is useless; the constant is outside the range of type '{0}'";
+
+        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterNodeAction(
+                CheckBinary,
+                SyntaxKind.GreaterThanExpression,
+                SyntaxKind.GreaterThanOrEqualExpression,
+                SyntaxKind.LessThanExpression,
+                SyntaxKind.LessThanOrEqualExpression);
+
+        private static void CheckBinary(SonarSyntaxNodeReportingContext context)
+        {
+            var binary = (BinaryExpressionSyntax)context.Node;
+            var typeIdentifier = string.Empty;
+            if (ShouldRaise(context.SemanticModel, binary.Left, binary.Right, ref typeIdentifier))
+            {
+                context.ReportIssue(Diagnostic.Create(Rule, binary.GetLocation(), typeIdentifier));
+            }
+
+            if (ShouldRaise(context.SemanticModel, binary.Right, binary.Left, ref typeIdentifier))
+            {
+                context.ReportIssue(Diagnostic.Create(Rule, binary.GetLocation(), typeIdentifier));
+            }
+        }
+
+        private static bool ShouldRaise(SemanticModel model, SyntaxNode expectedConstant, SyntaxNode other, ref string typeIdentifier) =>
+            TryGetConstant(model, expectedConstant, out var constant)
+            && model.GetSymbolInfo(other).Symbol.GetSymbolType() is { } symbol
+            && TryGetRange(symbol, out typeIdentifier, out var min, out var max)
+            && (constant < min || constant > max);
+
+        private static bool TryGetConstant(SemanticModel model, SyntaxNode expectedConstant, out double constant)
+        {
+            if (FindConstant(model, expectedConstant, Convert.ToDouble) is { } constValue)
+            {
+                constant = constValue;
+                return true;
+            }
+
+            constant = default;
+            return false;
+        }
```

Comment:
The problem with doing an if-else-if, is that on an example like the following:

```csharp
float x = 42;
const double y = double.MaxValue;
_ = x < y;
```

...the rule will not raise, because the Semantic Model resolves the `x` variable to the constant `42`, meaning it will go into the first `TryGetConstant` if-block and then it will never go to the `else-if` branch.

I think it's necessary to do if-should-raise-oneside -> if-should-raise-other-side instead of if-else-if, what do you think?

Reply:
Good question. I would say that this is another rule: "Don't compare constant values" because the result of the comparison behaves like a constant that is known at compile time (see also https://github.com/dotnet/csharplang/discussions/1028).

It fits into the "Silly mathematical comparisons should not be made" bucket (except that it is also useful for non-mathematical comparisons like `true == false` or `"Hello" == "World"`.

You could do a check before or after about both sides of the binary being a constant and raise S2198 but with another message like "Don't compare constant values".

What do you think?

Reply:
I implemented a different diagnostic message (the one you suggested). 

You can find the changes for this features on the last commit, along with some testcase changes.

This means that we get the constant values for both of the sides, but only do the weird math stuff when `exactly one` of them is actually a constant.

If none is a constant, we exit, if both are a constant, we raise your suggested message.